pkg/compactor/blocks_cleaner.go

```diff
@@ -411,11 +467,17 @@ func (c *BlocksCleaner) cleanUser(ctx context.Context, userID string, userLogger
 	}()
 
 	// Read the bucket index.
-	idx, err := bucketindex.ReadIndex(ctx, c.bucketClient, userID, c.cfgProvider, userLogger)
+
+	var idx *bucketindex.Index
+	err := withRetries(ctx, 1*time.Minute, c.retryConfig, log.With(userLogger, "op", "readIndex"), func(ctx context.Context) error {
+		var err error
+		idx, err = c.readIndex(ctx, c.bucketClient, userID, c.cfgProvider, userLogger)
```

Comment:
I'd prefer to just call `bucketindex.ReadIndex` here, and same for `c.writeIndex` vs `bucketindex.WriteIndex` later.

If we want to inject errors from read/write index calls, I'd suggest doing it at bucket level (see `ErrorInjectedBucketClient`, `deceivingUploadBucket` or `errBucket`), instead of introducing indirection in `BlocksCleaner`.